# TFG_AD_risk_factors
Genetic clues to AD: Uncovering Environmental Risk factors through GWAS

The project is prepared to be downloaded and run the codes from the "General" directory for which the pathways are prepared. Otherwise, just adjust the pathways of where the data is located and the other directories are. 

Since GWAS data files are too large the usage of Git Large File Storage (LFS) is needed. 

In order for the scripts to work, the following directories are needed:
| **Directory**          | **Content** | **Comments** |
|------------------------|------------|-------------|
| **Raw_Data/**          | GWAS summary statistics for each trait. | These files exceed GitHub's memory capacity and must be added separately. |
| **preprocessed_data/** | Contains four files per trait generated by the `preprocessing_data` script: <br> - `trait_effect_snp_clean.tsv` <br> - `trait_effect_snp_significant.tsv` <br> - `trait_snp_clean.tsv` <br> - `trait_snp_significant.tsv` | **Effect SNP files** (`*_effect_snp_*`): Include columns `rs_id`, `chr`, `pos`, `SE`, `beta`, and `p_value`. <br> **SNP files** (`*_snp_*`): Include columns `rs_id`, `chr`, `pos`, `effect_allele`, and `alt_allele`. |
| **processed_data/**    | Contains one file per trait, generated by the `processing_data` script. <br> File format: `AD1_trait2_filtered_snps.tsv` | These files represent the overlap between AD and environmental traits, after merging and distance-based filtering. Further filtering for significance can be applied. |
| **Images/**           | Contains visualization plots of variant distributions for each trait. | Includes one **Manhattan plot** and one **QQ plot** per trait. |

